### ISSUE TITLE
`try_err` suggests `return` inside eagerly evaluated `unwrap_or`

### ISSUE THREAD

**Issue by r00ster91** [rust-lang/rust-clippy#7199]
Lint name: `try_err`

I tried this code:

```rust
fn main() -> Result<(), &'static str> {
    Err(()).unwrap_or(Err("error")?);
    Ok(())
}
```

I expected to see this happen: No clippy warning or at least one that doesn't change the program's behavior.

Instead, this happened:

```
warning: returning an `Err(_)` with the `?` operator
 --> src/main.rs:2:23
  |
2 |     Err(()).unwrap_or(Err("error")?);
  |                       ^^^^^^^^^^^^^ help: try this: `return Err("error")`
  |
  = note: `#[warn(clippy::try_err)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#try_err

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 1.23s
```

If I apply the suggestion, this is the result:

```rust
fn main() -> Result<(), &'static str> {
    Err(()).unwrap_or(return Err("error"));
    Ok(())
}
```
`unwrap_or` is eagerly evaluated so the program behavior changed.

### Meta

- `cargo clippy -V`: clippy 0.1.52 (88f19c6 2021-05-03)
- `rustc -Vv`:
  ```
  rustc 1.52.0 (88f19c6da 2021-05-03)
  binary: rustc
  commit-hash: 88f19c6dab716c6281af7602e30f413e809c5974
  commit-date: 2021-05-03
  host: x86_64-unknown-linux-gnu
  release: 1.52.0
  LLVM version: 12.0.0
  ```

**Comment by hellow554** [comment]
I mean... this would work: `Err(()).unwrap_or({return Err("error")})` 😬

**Comment by HKalbasi** [comment]
`Err("error")?` is also eagerly evaluated. Your code is basically equal to a `return Err("error")`.

**Comment by r00ster91** [comment]
It seems you are right.
